File: doubly_linked_list.py
from node import Node

class DoublyLinkedList:
    def __init__(self):
        self.head = Node(0)  # head sentinel
        self.tail = Node(0)  # tail sentinel
        self.head.next = self.tail
        self.tail.prev = self.head

    def add_node(self, value):
        new_node = Node(value)
        last = self.tail.prev

        last.next = new_node
        new_node.prev = last
        new_node.next = self.tail
        self.tail.prev = new_node

    def print_list(self):
        current = self.head.next
        while current != self.tail:
            print(current.data, end=' ')
            current = current.next
        print()

File: DoublyLinkedList.java
public class DoublyLinkedList {

    private Node head; // head sentinel
    private Node tail; // tail sentinel

    public DoublyLinkedList() {
        head = new Node(0); // dummy value
        tail = new Node(0); // dummy value
        head.next = tail;
        tail.prev = head;
    }

    // Adds a node with given value to the end of the list
    public void addNode(int value) {
        Node newNode = new Node(value);
        Node last = tail.prev;

        last.next = newNode;
        newNode.prev = last;
        newNode.next = tail;
        tail.prev = newNode;
    }

    // Print list 
    public void printList() {
        Node current = head.next;
        while (current != tail) {
            System.out.print(current.data + " ");
            current = current.next;
        }
        System.out.println();
    }

}


